src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -167,15 +170,17 @@ async def mypy_typecheck(
         ),
         python_setup,
     )
-    use_subsystem_constraints = (
-        not mypy.options.is_default("interpreter_constraints")
-        or code_interpreter_constraints.includes_python2()
-    )
-    tool_interpreter_constraints = (
-        PexInterpreterConstraints(mypy.interpreter_constraints)
-        if use_subsystem_constraints
-        else code_interpreter_constraints
-    )
+
+    if not mypy.options.is_default("interpreter_constraints"):
+        tool_interpreter_constraints = mypy.interpreter_constraints
+    elif code_interpreter_constraints.requires_python38_or_newer():
+        tool_interpreter_constraints = ("CPython>=3.8",)
+    elif code_interpreter_constraints.requires_python37_or_newer():
+        tool_interpreter_constraints = ("CPython>=3.7",)
+    elif code_interpreter_constraints.requires_python36_or_newer():
+        tool_interpreter_constraints = ("CPython>=3.6",)
+    else:
+        tool_interpreter_constraints = mypy.interpreter_constraints
```

Comment:
This will get simplified soon with batching by interpreter constraints, because we will start to set MyPy's `python_version` ourselves. 

For now, we have it this way to 

a) ensure Py38 code uses Py38 to run MyPy, which is necessary to work properly, and
b) you don't have to specify `python_version` in `mypy.ini` because Pants will run with the correct Python interpreter, which is used to set the default

See https://github.com/pantsbuild/pants/pull/10750 for more about this.